### PR TITLE
WIP: RR graph uxsdcxx based reader and writer

### PR DESCRIPTION
**Pull request by litghost**
Changes rr graph reader to be based on the interface oriented version of uxsdxx.  Performance of  the XML is improved (25 to 20 seconds), and the capnp based reader is faster (25 to 10 seconds), and half the memory!  See #333 

This PR is for code review of `vpr/src/route/rr_graph_uxsdcxx_interface_impl.h` and to run the CI.

### CODE REVIEW

**Review by acomodi — APPROVED** [review]
LGTM.
Just to know, the last two commits are not part of any wip branches right? If so I wonder if it would be better to include them in the fast_rr_read wip branch.

**Comment by litghost** [comment]
The current structure is intentional.  I wanted to kick off a ci run asap.

More work is required before this can be merged.

**Comment by litghost** [comment]
VtR + Symbiflow failed because of https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1200.  Once https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1200 is merged, I'll re-run CI and hopefully it will be green.

**Review by HackerFoo — CHANGES_REQUESTED** [review]
There are a lot of generated files. Can we add the generator to the build process instead?

**Comment by litghost** [comment]
> There are a lot of generated files. Can we add the generator to the build process instead?

VTR has no python infrastructure, and a base python instance is not sufficient for the generator.

**Comment by litghost** [comment]
> There are a lot of generated files. Can we add the generator to the build process instead?

To be clear, this was actually discussed with the VTR devs, and they want to check in the generated files, along with a path (e.g. a CMake target) to regenerate the files.

File: vpr/src/route/rr_graph_uxsdcxx_interface_impl.h [hunk 1]
```diff
@@ -0,0 +1,1919 @@
...
+                std::string,
+                std::vector<t_metadata_value>>*>(data);
+        VTR_ASSERT(meta_value->second.size() == 1);
+        return meta_value->second[0].as_string().c_str();
+    }
+    inline size_t num_metadata_meta(void* data) override {
+        auto* meta = static_cast<struct t_metadata_dict*>(data);
+        return meta->size();
+    }
+    inline void* get_metadata_meta(int n, void* data) override {
```
Comment (duck2) [inline vpr/src/route/rr_graph_uxsdcxx_interface_impl.h]: Maybe it's better to generate a `begin/next/end` set of functions in uxsdcxx? We would need one `*data` for the iterator state and one `*data` for the parent state.

File: vpr/src/route/rr_graph_uxsdcxx_interface_impl.h [hunk 2]
```diff
@@ -0,0 +1,1919 @@
...
+        return uxsd::enum_node_type::UXSD_INVALID;
+    }
+    inline void* get_node_loc(void* data) override {
+        return data;
+    }
+    inline void* get_node_timing(void* data) override {
+        return data;
+    }
+    inline bool has_node_timing(void* /*data*/) override {
+        return true;
```
Comment (duck2) [inline vpr/src/route/rr_graph_uxsdcxx_interface_impl.h]: This can be fixed in the schema.

**Review by HackerFoo — APPROVED** [review]
> > There are a lot of generated files. Can we add the generator to the build process instead?
> 
> To be clear, this was actually discussed with the VTR devs, and they want to check in the generated files, along with a path (e.g. a CMake target) to regenerate the files.

I'm okay with checking in generated files (although it's not ideal) as long as I can easily regenerate them locally.

**Comment by litghost** [comment]
Superceeded by #365

**Pull request closed without merging**